### ISSUE TITLE
Sparsity los for get_jacobian in Callback

### ISSUE THREAD

**Issue by jgillis** [casadi/casadi#1847]
``` python
from casadi import *
import numpy as np

class Convolution(Callback):
  def __init__(self,A, flag_transp):
    Callback.__init__(self)
    self.A = A
    if flag_transp:
      self.m,self.n = A.shape
    else:
      self.n,self.m = A.shape
    self.data = []
    self.flag_transp = flag_transp
    self.construct('Convolution')
  def get_transpose(self):
    out = Convolution(self.A,not self.flag_transp)
    self.data.append(out)
    return out
  def get_sparsity_in(self,i):
    return casadi.Sparsity.dense(self.m,1)

  def get_sparsity_out(self,i):
    return casadi.Sparsity.dense(self.n,1)

  def eval(self, ins):
    if self.flag_transp:
        y = mtimes(self.A.T,x[0])
    else:
        y = mtimes(self.A,x[0])
    return [y]
  def has_jacobian(self):
    return True
  def get_jacobian(self,name,opts):
    x = casadi.MX.sym("x",self.n,1)
    f = casadi.Function('f',[x],[casadi.diag(x)])
    print f
    return f


x = MX.sym("x",5)
A= np.random.rand(5,5)

F = Convolution(A,False)

print F.jacobian()

g = Function("F",[x],[F(x*5)])

print g.jacobian()

```

**Comment by jgillis** [comment]
Quickfix:

```python
F.set_jac_sparsity(casadi.Sparsity.diag(5),0,0,True)
```

**Comment by jaeandersson** [comment]
I'll try to make a more proper fix, by refactoring `FunctionInternal::sparsity_jac`.

**Comment by jgillis** [comment]
Quickfix reported not to work properly

**Comment by jaeandersson** [comment]
OK, I'll bump priority. Bug me if I don't fix it...

**Comment by jgillis** [comment]
bump @jaeandersson

**Comment by jgillis** [comment]
bump @jaeandersson

**Comment by jaeandersson** [comment]
Ah, the refactoring of this is only halfway through. Should finish that..

**Comment by jgillis** [comment]
ping @jaeandersson

**Comment by jaeandersson** [comment]
Note: One of the first things needed is to replace:
```c++
    /** \brief Get Jacobian sparsity */
    virtual Sparsity get_jacobian_sparsity() const;
```
with a cached
```c++
    ///@{
    /** \brief Get Jacobian sparsity */
    Sparsity jacobian_sparsity() const;
    virtual bool has_jacobian_sparsity() const { return false;}
    virtual Sparsity get_jacobian_sparsity() const;
    ///@}
```

**Comment by jaeandersson** [comment]
`get_jacobian_sparsity` would need to be implemented using seed propagation.

**Comment by jgillis** [comment]
From earlier private conversation:
```
You'd want to add the option for FunctionInternal to have a cache for the sparsity pattern of all the Jacobian blocks. This would correspond to the signature of FunctionInternal::jac, which should completely replace FunctionInternal::jacobian
to do it properly, you have to refactor the (hierarchical) jacobian sparsity pattern calculation to allow calculation of multiple blocks at once, efficiently.
code is not very clean at the moment, that's a problem
```

**Comment by chrisdembia** [comment]
The [OpenSim Moco project](https://github.com/opensim-org/opensim-moco) is depending on the `has_jacobian_sparsity()`/`get_jacobian_sparsity()` feature. Is it possible to get this merged into an official release so we can depend on an official release?

**Comment by jgillis** [comment]
Now merged into 3.5.2

**Comment by chrisdembia** [comment]
Yay! Thank you :)
